docs(snowflake): add blog post showing insertion into snowflake from postgres (#8426)
ibis-project/ibis: docs/posts/into-snowflake/index.qmd

---
title: "Snow IO: loading data from other DBs into Snowflake"
author: "Phillip Cloud"
error: false
date: "2024-02-22"
categories:
  - blog
  - snowflake
  - io
  - productivity
---

## Recap

We've [blogged about Snowflake IO before](../snowflake-io/index.qmd), in the
context of getting local files into Snowflake as fast as possible.

> In this post, we'll show how query another system and insert query results into

Reviewer: ```suggestion
In this post, we'll show how to query another system and insert query results into
```

We'll connect to a postgres database running locally in a container. You
should be able to swap in your own connection details as needed.

```{python}
from ibis.interactive import *  # <1>

pg_con = ibis.connect("postgres://postgres:postgres@localhost/postgres")
```

1. Import Ibis for maximum productivity in interactive analysis.

We'll use a test dataset that contains some baseball batting statistics.

Ibis provides that example data, so we can dump that into postgres.

```{python}
pg_batting = pg_con.create_table(
    "batting",
    ibis.examples.Batting.fetch().to_pandas(),  # <1>
    temp=True,  # <2>
)
```

1. Yep, I'm using pandas here!
2. Use a temporary table to avoid cluttering up the database.

### Connect to Snowflake

```{python}
import os

snow_con = ibis.connect(os.environ["SNOWFLAKE_URL"])  # <1>
```

1. Set the `SNOWFLAKE_URL` environment variable to your Snowflake connection
   string, which will look like `snowflake://user:pass@account/database/schema?warehouse=my_warehouse`.

## Profit

### Construct an Ibis expression from the postgres data

Let's build an Ibis expression based on the `batting` table in our postgres database.

```{python}
pg_batting
```

> Let's compute the average RBI per year per team.

Reviewer: ```suggestion
We can compute the average RBI per year per team.
```

or similar, just to break up the "let's, let's, let's"

Reviewer 2: Can we add either a link to what RBI is or a "(Runs Batted In)" for people that might have no clue about baseball.

Author: Definitely. Link incoming.

```{python}
pg_expr = pg_batting.group_by(("year_id", "team_id")).agg(avg_rbi=_.rbi.mean())
pg_expr
```

Let's show how many rows we have in the result.

```{python}
pg_expr.count()
```

### Insert the computed results into Snowflake

Because all of our backends implement the `to_pyarrow()` method, we can
get data out of another system and into Snowflake with a few lines of code.

First we'll create a table in Snowflake to hold the data. Ibis helps here by
providing an API to access the schema from the **postgres**-based expression,
so you don't need to worry about translating between the postgres type system

> and the Snowflake type system.

Reviewer: Maybe here we spell out more directly that "...Ibis translates from the postgres type system to the Snowflake type system so you don't have to"?

Author: Done

```{python}
snow_table = snow_con.create_table("pg_batting", schema=pg_expr.schema(), temp=True)  # <1>
```

1. By default the table will be created in the database and schema of the
   current connection.

   We create a temporary table for the same reason we do with postgres above.

```{python}
snow_con.insert("pg_batting", pg_expr.to_pyarrow())
```

To sanity check what we've done let's peek at the table.

```{python}
snow_table
```

We'll count them too, to be extra sure.

```{python}
snow_table.count()
```

## Conclusion

> In this post we show how easy it is to move data from one backend into Snowflake.

Reviewer 2: It's easy with Ibis, maybe mentioned that? 

```suggestion
In this post we show how easy it is to move data from one backend into Snowflake using Ibis.
```

We'll connect to a postgres database running locally in a container. You
should be able to swap in your own connection details as needed.

```{python}
from ibis.interactive import *  # <1>

pg_con = ibis.connect("postgres://postgres:postgres@localhost/postgres")
```

1. Import Ibis for maximum productivity in interactive analysis.

We'll use a test dataset that contains some baseball batting statistics.

Ibis provides that example data, so we can dump that into postgres.

```{python}
pg_batting = pg_con.create_table(
    "batting",
    ibis.examples.Batting.fetch().to_pandas(),  # <1>
    temp=True,  # <2>
)
```

1. Yep, I'm using pandas here!
2. Use a temporary table to avoid cluttering up the database.

### Connect to Snowflake

```{python}
import os

snow_con = ibis.connect(os.environ["SNOWFLAKE_URL"])  # <1>
```

1. Set the `SNOWFLAKE_URL` environment variable to your Snowflake connection
   string, which will look like `snowflake://user:pass@account/database/schema?warehouse=my_warehouse`.

## Profit

### Construct an Ibis expression from the postgres data

Let's build an Ibis expression based on the `batting` table in our postgres database.

```{python}
pg_batting
```

> We can compute the average RBI per year per team.

Reviewer 2: Can we add either a link to what RBI is or a "(Runs Batted In)" for people that might have no clue about baseball.

---
title: "Snow IO: loading data from other DBs into Snowflake"
author: "Phillip Cloud"
error: false

> date: "2024-02-23"

Reviewer 3: can you update the date to today, then I'll merge? I don't have snowflake setup or I'd do it myself

Author: Done

categories:
  - blog
  - snowflake
  - io
  - productivity
---

## Recap

We've [blogged about Snowflake IO before](../snowflake-io/index.qmd), in the
context of getting local files into Snowflake as fast as possible.

In this post, we'll show how to insert query results from another system into
Snowflake, using Ibis.

## Setup

### Connect to your non-Snowflake system

We'll connect to a postgres database running locally in a container. You

> should be able to swap in your own connection details as needed.

Reviewer 3: would it be worth mentioning `just up postgres` if they want to try this themselves? of course in practice they'll have their own database somewhere

```{python}
from ibis.interactive import *  # <1>

pg_con = ibis.connect("postgres://postgres:postgres@localhost/postgres")
```

1. Import Ibis for maximum productivity in interactive analysis.

We'll use a test dataset that contains some baseball batting statistics.

Ibis provides that example data, so we can dump that into postgres.

```{python}
pg_batting = pg_con.create_table(
    "batting",

>     ibis.examples.Batting.fetch().to_pandas(),  # <1>

Reviewer 3: why pandas instead of pyarrow?

Author: psycopg2 doesn't seem to be buying what we're selling with `to_pyarrow()` :)

I'll open an issue about it.

    temp=True,  # <2>
)
```

1. Yep, I'm using pandas here!
2. Use a temporary table to avoid cluttering up the database.

### Connect to Snowflake

```{python}
import os

snow_con = ibis.connect(os.environ["SNOWFLAKE_URL"])  # <1>
```

1. Set the `SNOWFLAKE_URL` environment variable to your Snowflake connection

>    string, which will look like `snowflake://user:pass@account/database/schema?warehouse=my_warehouse`.

Reviewer 3: doesn't really matter but looks like an extra space on the second line

also, given how long this code tip is, perhaps consider a callout block above?

Reviewer 3: actually it's more that this doesn't look good when hovered over -- the long code tip below is fine -- because of the long URI here you have to scroll right while hovering over

<img width="347" alt="image" src="https://github.com/ibis-project/ibis/assets/54814569/0544bb99-a065-4a5f-a907-fa29a143119f">

## Profit

### Construct an Ibis expression from the postgres data

Let's build an Ibis expression based on the `batting` table in our postgres database.

```{python}
pg_batting
```

We can compute the average [RBI](https://en.wikipedia.org/wiki/Run_batted_in) per year per team.

```{python}
pg_expr = pg_batting.group_by(("year_id", "team_id")).agg(avg_rbi=_.rbi.mean())
pg_expr
```

We can also rename columns to be more consistent with typical Snowflake usage.

```{python}
pg_expr = pg_expr.rename("ALL_CAPS")
pg_expr
```

Let's show how many rows we have in the result.

```{python}
pg_expr.count()
```

### Insert the computed results into Snowflake

> Because all of our backends implement the `to_pyarrow()` method, we can

Reviewer 3: nit: "of our" -> "Ibis"

get data out of another system and into Snowflake with a few lines of code.

First we'll create a table in Snowflake to hold the data.

Ibis helps here by providing an API to access the schema from the
**postgres**-based expression, and automatically translates postgres types into
Snowflake types.

```{python}
snow_table = snow_con.create_table("pg_batting", schema=pg_expr.schema(), temp=True)  # <1>
```

1. By default the table will be created in the database and schema of the
   current connection.

   We create a temporary table for the same reason we do with postgres above.

We'll show that the table is empty to sanity check ourselves.

```{python}
snow_table
```

Insert the expression's result table into Snowflake.

```{python}
snow_con.insert("pg_batting", pg_expr.to_pyarrow())
```

To sanity check what we've done let's peek at the table.

```{python}
snow_table
```

We'll count them too, to be extra sure.

```{python}
snow_table.count()
```

## Conclusion

> In this post we show how easy it is to move data from one backend into Snowflake using Ibis.

Reviewer 4: I love the blog post because I think this is an important use case to move data from transactional data stores into an OLAP system. The blog provides an effective way for the batch move. There is a subset of users who will need to keep the OLTP and OLAP in sync with minimal latency, which will require CDC support (we don't have the support yet). It'll be a good idea to allude to this capability and get feedback from the community for this need. Here is a good read to keep MySQL and Iceberg in sync. https://ververica.github.io/flink-cdc-connectors/master/content/quickstart/build-real-time-data-lake-tutorial.html

Reviewer 4: completely reasonable
